diff --git a/include/circle/usb/gadget/dwusbgadget.h b/include/circle/usb/gadget/dwusbgadget.h
index 1c9cbf43..ac84e75b 100644
--- a/include/circle/usb/gadget/dwusbgadget.h
+++ b/include/circle/usb/gadget/dwusbgadget.h
@@ -39,7 +39,8 @@ public:
 public:
 	/// \param pInterruptSystem Pointer to the interrupt system
 	/// \param DeviceSpeed Enumeration speed of this USB gadget
-	CDWUSBGadget (CInterruptSystem *pInterruptSystem, TDeviceSpeed DeviceSpeed);
+	CDWUSBGadget (CInterruptSystem *pInterruptSystem, TDeviceSpeed DeviceSpeed,
+		u16 usVendorId = USB_GADGET_VENDOR_ID, u16 usProductId = USB_GADGET_DEVICE_ID_BASE);
 
 	virtual ~CDWUSBGadget (void);
 
@@ -74,6 +75,8 @@ public:
 	/// \return Number of bytes in pData to be sent, or < 0 on error (STALL the request)
 	/// \note By default host-to-device requests will be ignored and other requests will be STALLed
 	virtual int OnClassOrVendorRequest (const TSetupData *pSetupData, u8 *pData);
+	u16 GetVendorId() const { return m_usVendorId; }
+	u16 GetProductId() const { return m_usProductId; }
 
 private:
 	boolean PowerOn (void);
@@ -108,6 +111,8 @@ private:
 	static const unsigned NumberOfEPs	= 7;	// without EP0
 	static const unsigned NumberOfInEPs	= 7;
 	static const unsigned NumberOfOutEPs	= 7;
+	u16 m_usVendorId;
+	u16 m_usProductId;
 
 private:
 	CInterruptSystem *m_pInterruptSystem;
@@ -136,4 +141,4 @@ private:
 	CDWUSBGadgetEndpoint *m_pEP[NumberOfEPs+1];
 };
 
-#endif
+#endif
\ No newline at end of file
diff --git a/lib/usb/gadget/dwusbgadget.cpp b/lib/usb/gadget/dwusbgadget.cpp
index dc3f93cd..360d6005 100644
--- a/lib/usb/gadget/dwusbgadget.cpp
+++ b/lib/usb/gadget/dwusbgadget.cpp
@@ -39,8 +39,11 @@
 
 LOGMODULE ("dwgadget");
 
-CDWUSBGadget::CDWUSBGadget (CInterruptSystem *pInterruptSystem, TDeviceSpeed DeviceSpeed)
-:	m_pInterruptSystem (pInterruptSystem),
+CDWUSBGadget::CDWUSBGadget (CInterruptSystem *pInterruptSystem, TDeviceSpeed DeviceSpeed,
+                            u16 usVendorId, u16 usProductId)
+:	m_usVendorId (usVendorId),
+	m_usProductId (usProductId),
+	m_pInterruptSystem (pInterruptSystem),
 	m_DeviceSpeed (DeviceSpeed),
 	m_State (StatePowered),
 	m_bPnPEvent {FALSE, FALSE}
