diff --git a/include/circle/usb/dwhciregister.h b/include/circle/usb/dwhciregister.h
index bff54a95..50463618 100644
--- a/include/circle/usb/dwhciregister.h
+++ b/include/circle/usb/dwhciregister.h
@@ -23,6 +23,9 @@
 #include <circle/usb/dwhci.h>
 #include <circle/types.h>
 
+#define DWHCI_DEV_STAT                  (ARM_USB_BASE + 0x808)
+#define DWHCI_DEV_STAT_ENUM_SPD__MASK   0x00000006
+#define DWHCI_DEV_STAT_ENUM_SPD__SHIFT  1
 class CDWHCIRegister
 {
 public:
diff --git a/include/circle/usb/gadget/dwusbgadget.h b/include/circle/usb/gadget/dwusbgadget.h
index 1c9cbf43..0db8f762 100644
--- a/include/circle/usb/gadget/dwusbgadget.h
+++ b/include/circle/usb/gadget/dwusbgadget.h
@@ -74,6 +74,8 @@ public:
 	/// \return Number of bytes in pData to be sent, or < 0 on error (STALL the request)
 	/// \note By default host-to-device requests will be ignored and other requests will be STALLed
 	virtual int OnClassOrVendorRequest (const TSetupData *pSetupData, u8 *pData);
+	/// \return Negotiated USB speed after enumeration
+	TDeviceSpeed GetNegotiatedUSBSpeed (void) const { return m_NegotiatedSpeed; }
 
 private:
 	boolean PowerOn (void);
@@ -112,6 +114,7 @@ private:
 private:
 	CInterruptSystem *m_pInterruptSystem;
 	TDeviceSpeed m_DeviceSpeed;
+	TDeviceSpeed m_NegotiatedSpeed;
 
 	enum TState
 	{
